rust-sdk/core/src/quote/swap.rs

```diff
@@ -0,0 +1,677 @@
+    TickArraySequence, TickFacade, TransferFee, WhirlpoolFacade, U128,
+};
+
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::*;
+
+const MIN_SQRT_PRICE: u128 = 4295048016;
+const MAX_SQRT_PRICE: u128 = 79226673515401279992447579055;
+
+/// Computes the exact input or output amount for a swap transaction.
+///
+/// # Arguments
+/// - `token_in`: The input token amount.
+/// - `specified_token_a`: If `true`, the input token is token A. Otherwise, it is token B.
+/// - `slippage_tolerance`: The slippage tolerance in basis points.
+/// - `whirlpool`: The whirlpool state.
+/// - `tick_array_0`: The tick array at the current tick index.
+/// - `tick_array_plus_1`: The tick array at the current tick offset plus 1.
+/// - `tick_array_plus_2`: The tick array at the current tick offset plus 2.
+/// - `tick_array_minus_1`: The tick array at the current tick offset minus 1.
+/// - `tick_array_minus_2`: The tick array at the current tick offset minus 2.
+/// - `transfer_fee_a`: The transfer fee for token A.
+/// - `transfer_fee_b`: The transfer fee for token B.
+///
+/// # Returns
+/// The exact input or output amount for the swap transaction.
+#[allow(clippy::too_many_arguments)]
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = swapQuoteByInputToken, skip_jsdoc))]
+pub fn swap_quote_by_input_token(
+    token_in: U128,
+    specified_token_a: bool,
+    slippage_tolerance: u16,
+    whirlpool: WhirlpoolFacade,
+    tick_array_0: TickArrayFacade,
+    tick_array_plus_1: TickArrayFacade,
+    tick_array_plus_2: TickArrayFacade,
+    tick_array_minus_1: TickArrayFacade,
+    tick_array_minus_2: TickArrayFacade,
+    transfer_fee_a: Option<TransferFee>,
+    transfer_fee_b: Option<TransferFee>,
+) -> ExactInSwapQuote {
+    let (transfer_fee_in, transfer_fee_out) = if specified_token_a {
+        (transfer_fee_b, transfer_fee_a)
+    } else {
+        (transfer_fee_a, transfer_fee_b)
+    };
+    let token_in_after_fee = adjust_amount(token_in.into(), transfer_fee_in.into(), false);
+
+    let tick_sequence = TickArraySequence::new(
+        [
+            tick_array_minus_2,
+            tick_array_minus_1,
+            tick_array_0,
+            tick_array_plus_1,
+            tick_array_plus_2,
+        ],
+        whirlpool.tick_spacing,
+    );
+
+    let swap_result = compute_swap(
+        token_in_after_fee.into(),
+        whirlpool,
+        tick_sequence,
+        specified_token_a,
+        true,
+    );
+
+    let (token_in_after_fees, token_est_out_before_fee) = if specified_token_a {
+        (swap_result.token_a, swap_result.token_b)
+    } else {
+        (swap_result.token_b, swap_result.token_a)
+    };
+
+    let token_min_out_before_fee = adjust_amount(
+        token_est_out_before_fee.into(),
+        AdjustmentType::Slippage(slippage_tolerance),
+        false,
+    );
+
+    let token_in = inverse_adjust_amount(token_in_after_fees.into(), transfer_fee_in.into(), false);
```

Comment:
`Smaller amount will be chosen` so in that case inverse_adjust_amount(99, 1%) will result 100. IMO this might be desired because we would want the user to spend the least amount to get the maximum output. i.e. whether a suer spends `101` or `100` the output amount would be the same I guess